applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_cr_beam_element_3D2N.cpp

```diff
@@ -0,0 +1,287 @@
+}
+
+void AdjointFiniteDifferenceCrBeamElement::Calculate(const Variable<Vector >& rVariable,
+                        Vector& rOutput,
+                        const ProcessInfo& rCurrentProcessInfo)
+{
+    KRATOS_TRY;
+
+    if(rVariable == STRESS_ON_GP || rVariable == STRESS_ON_NODE)
+    {
+        TracedStressType traced_stress_type = static_cast<TracedStressType>(this->GetValue(TRACED_STRESS_TYPE));
+
+        std::vector< array_1d<double, 3 > > stress_vector;
+        int direction_1 = 0;
+        bool stress_is_moment = true;
+
+        switch (traced_stress_type)
+        {
+            case TracedStressType::MX:
+            {
+                direction_1 = 0;
+                break;
+            }
+            case TracedStressType::MY:
+            {
+                direction_1 = 1;
+                break;
+            }
+            case TracedStressType::MZ:
+            {
+                direction_1 = 2;
+                break;
+            }
+            case TracedStressType::FX:
+            {
+                direction_1 = 0;
+                stress_is_moment = false;
+                break;
+            }
+            case TracedStressType::FY:
+            {
+                direction_1 = 1;
+                stress_is_moment = false;
+                break;
+            }
+            case TracedStressType::FZ:
+            {
+                direction_1 = 2;
+                stress_is_moment = false;
+                break;
+            }
+            default:
+                KRATOS_ERROR << "Invalid stress type! Stress type not supported for this element!" << std::endl;
+        }
+
+        if(stress_is_moment)
+            mpPrimalElement->GetValueOnIntegrationPoints(MOMENT, stress_vector, rCurrentProcessInfo);
+        else
+            mpPrimalElement->GetValueOnIntegrationPoints(FORCE, stress_vector, rCurrentProcessInfo);
+
+        if(rVariable == STRESS_ON_GP)
+        {
+            const unsigned int&  GP_num = GetGeometry().IntegrationPointsNumber(Kratos::GeometryData::GI_GAUSS_3);
+
+            rOutput.resize(GP_num);
+            for(unsigned int i = 0; i < GP_num ; i++)
+            {
+                rOutput(i) = stress_vector[i][direction_1];
+            }
+        }
+        else if(rVariable == STRESS_ON_NODE)
+        {
+            rOutput.resize(2);
+            rOutput(0) = 2 * stress_vector[0][direction_1] - stress_vector[1][direction_1];
+            rOutput(1) = 2 * stress_vector[2][direction_1] - stress_vector[1][direction_1];
+        }
+    }
+    else
+    {
+        rOutput.resize(3);
+        rOutput.clear();
+    }
+
+    KRATOS_CATCH("")
+}
+
+int AdjointFiniteDifferenceCrBeamElement::Check(const ProcessInfo& rCurrentProcessInfo)
+{
+    KRATOS_TRY
+
+    KRATOS_ERROR_IF(GetGeometry().WorkingSpaceDimension() != 3 || GetGeometry().size() != 2)
+    << "The beam element works only in 3D and with 2 noded elements" << "" << std::endl;
+
+    KRATOS_CHECK_VARIABLE_KEY(VELOCITY);
```

Comment:
Use the base class Check.

Reply:
Not possible because of DOF check. Or do you mean the base class of the adjoint element?